src/Microsoft.ML.CpuMath/FactorizationMachine/AvxIntrinsics.cs

```diff
@@ -0,0 +1,188 @@
+            Sse.StoreScalar(&latentResponse, y.GetLower()); // The lowest slot is the response value.
+            *response = linearResponse + latentResponse;
+        }
+
+        // This function implements Algorithm 2 in https://github.com/wschin/fast-ffm/blob/master/fast-ffm.pdf
+        // Calculate the stochastic gradient and update the model.
+        public static unsafe void CalculateGradientAndUpdate(float lambdaLinear, float lambdaLatent, float learningRate, int fieldCount, int latentDim, float weight, int count,
+            int* fieldIndices, int* featureIndices, float* featureValues, float* latentSum, float slope, float* linearWeights, float* latentWeights,
+            float* linearAccumulatedSquaredGrads, float* latentAccumulatedSquaredGrads)
+        {
+            int m = fieldCount;
+            int d = latentDim;
+            int c = count;
+            int* pf = fieldIndices;
+            int* pi = featureIndices;
+            float* px = featureValues;
+            float* pq = latentSum;
+            float* pw = linearWeights;
+            float* pv = latentWeights;
+            float* phw = linearAccumulatedSquaredGrads;
+            float* phv = latentAccumulatedSquaredGrads;
+
+            Vector256<float> wei = Vector256.Create(weight);
+            Vector256<float> s= Vector256.Create(slope);
+            Vector256<float> lr = Vector256.Create(learningRate);
+            Vector256<float> lambdav = Vector256.Create(lambdaLatent);
+
+            for (int i = 0; i < count; i++)
+            {
+                int f = pf[i];
+                int j = pi[i];
+
+                // Calculate gradient of linear term w_j.
+                float g = weight * (lambdaLinear * pw[j] + slope * px[i]);
+
+                // Accumulate the gradient of the linear term.
+                phw[j] += g * g;
+
+                // Perform ADAGRAD update rule to adjust linear term.
+                pw[j] -= learningRate / MathF.Sqrt(phw[j]) * g;
+
+                // Update latent term, v_j,f', f'=1,...,m.
+                Vector256<float> x = Avx.BroadcastScalarToVector256(px + i);
+
+                for (int fprime = 0; fprime < m; fprime++)
+                {
+                    float* vjfprime = pv + j * m * d + fprime * d;
+                    float* hvjfprime = phv + j * m * d + fprime * d;
+                    float* qfprimef = pq + fprime * m * d + f * d;
+                    Vector256<float> sx = Avx.Multiply(s, x);
+
+                    for (int k = 0; k + 8 <= d; k += 8)
+                    {
+                        Vector256<float> v = Avx.LoadVector256(vjfprime + k);
+                        Vector256<float> q = Avx.LoadVector256(qfprimef + k);
+
+                        // Calculate L2-norm regularization's gradient.
+                        Vector256<float> gLatent = Avx.Multiply(lambdav, v);
+
+                        // Calculate loss function's gradient.
+                        if (fprime != f)
```

Comment:
is this better than creating a temporary to hold the multiplier and then doing `gLatent = Avx.Add(gLatent, Avx.Multiply(sx, tmp))` outside the if/else?

Reply:
I tried this as suggested and I see a similar performance.